File: src/ts-engine/pack.ts
/** Token-budget packer — port of pack.py */

import { createHash } from "node:crypto";
import { jaccard, keywordSet } from "./extractive.ts";
import { estimateTokens } from "./metrics.ts";

export const DEFAULT_FORWARD_BUDGET = 2048;
export const DEDUP_K = 3;
export const WARMUP_TURNS = 3;
export const MARGINAL_JACCARD = 0.92;
export const SKIP_FLOOR_TOKENS = 64;
export const NOVELTY_BUDGET_FLOOR = 1.0;

export function forwardBudget(): number {
  const raw = (process.env.CHAT_COMPRESSOR_FORWARD_BUDGET || "").trim();
  if (!raw) return DEFAULT_FORWARD_BUDGET;
  const n = Number.parseInt(raw, 10);
  return Number.isFinite(n) && n >= 1 ? n : DEFAULT_FORWARD_BUDGET;
}

export function noveltyBudgetFloor(): number {
  const raw = (process.env.CHAT_COMPRESSOR_NOVELTY_FLOOR || "").trim();
  if (!raw) return NOVELTY_BUDGET_FLOOR;
  const v = Number.parseFloat(raw);
  if (!Number.isFinite(v)) return NOVELTY_BUDGET_FLOOR;
  return Math.max(0, Math.min(1, v));
}

export function marginalJaccardThreshold(defaultVal = MARGINAL_JACCARD): number {
  const raw = (process.env.CHAT_COMPRESSOR_MARGINAL_JACCARD || "").trim();
  if (!raw) return defaultVal;
  const v = Number.parseFloat(raw);
  if (!Number.isFinite(v)) return defaultVal;
  return Math.max(0, Math.min(1, v));
}

export function crossTurnDedupEnabled(): boolean {
  const raw = (process.env.CHAT_COMPRESSOR_CROSS_TURN_DEDUP || "1").trim().toLowerCase();
  return !["0", "false", "no", "off"].includes(raw);
}

export function countTokens(text: string): { ws: number; chars4: number } {
  const ws = text.trim() ? text.split(/\s+/).length : 0;
  const chars4 = text ? estimateTokens(text) : 0;
  return { ws, chars4 };
}

export function lineHash(text: string): string {
  return createHash("sha1")
    .update(text.trim().toLowerCase(), "utf8")
    .digest("hex")
    .slice(0, 16);
}

export function adaptiveBudget(
  t: number,
  noveltyRate: number,
  cap?: number,
  floor?: number,
): number {
  let limit = Math.max(1, cap ?? forwardBudget());
  if (t <= WARMUP_TURNS) return limit;
  const rate = Math.max(0, Math.min(1, noveltyRate));
  const rhoMin = floor ?? noveltyBudgetFloor();
  const scaled = Math.floor(limit * Math.max(rhoMin, rate));
  return Math.max(SKIP_FLOOR_TOKENS, Math.min(limit, scaled));
}

export type PackResult = {
  text: string;
  packed_tokens: number;
  tokens_ws: number;
  tokens_chars4: number;
  budget: number;
  rate: number;
  method: string;
  novel_tokens: number;
  dup_suppressed_tokens: number;
  line_hashes: string[];
  tau_hot: number;
  tau_typed: number;
  tau_ranked: number;
  tau_spans: number;
};

export function packForward(args: {
  hot_set?: string;
  typed_lines?: string[];
  ranked_chunks?: string[];
  span_chunks?: string[];
  budget?: number;
  recent_hashes?: Set<string>;
  openitem_changed?: boolean;
  node_superseded?: boolean;
  allow_skip?: boolean;
  marginal_jaccard?: number;
  skip_floor_tokens?: number;
}): PackResult {
  const cap = Math.max(1, args.budget ?? forwardBudget());
  const mu =
    args.marginal_jaccard === undefined
      ? marginalJaccardThreshold()
      : args.marginal_jaccard;
  const skipFloor = args.skip_floor_tokens ?? SKIP_FLOOR_TOKENS;
  const parts: string[] = [];
  let used = 0;
  const packedNorm = new Set<string>();
  let method = "hot_set";
  let dupSuppressed = 0;
  let tauHot = 0;
  let tauTyped = 0;
  let tauRanked = 0;
  let tauSpans = 0;
  let suppress = new Set(args.recent_hashes ?? []);
  if (args.node_superseded || !crossTurnDedupEnabled()) suppress = new Set();

  const blocked = (text: string) => (suppress.size ? suppress.has(lineHash(text)) : false);
  const extraParts: string[] = [];
  const marginal = (text: string) => {
    if (!extraParts.length) return false;
    return jaccard(keywordSet(text), keywordSet(extraParts.join("\n"))) > mu;
  };

  const hot = (args.hot_set || "").trim();
  if (hot) {
    let piece = `HOT_SET:\n${hot}`;
    let toks = estimateTokens(piece);
    if (toks > cap) {
      piece = truncateToBudget(piece, cap);
      toks = estimateTokens(piece);
    }
    parts.push(piece);
    used += toks;
    tauHot = toks;
    packedNorm.add(hot.toLowerCase());
    for (const line of hot.split("\n")) packedNorm.add(line.trim().toLowerCase());
  }

  const fits = (block: string) => used + estimateTokens(block) <= cap;

  let typedAdded = false;
  for (const line of args.typed_lines ?? []) {
    const text = (line || "").trim();
    if (!text) continue;
    const key = text.toLowerCase();
    if (packedNorm.has(key)) continue;
    if (blocked(text)) {
      dupSuppressed += estimateTokens(text);
      continue;
    }
    if (marginal(text)) {
      dupSuppressed += estimateTokens(text);
      continue;
    }
    if (!fits(text)) continue;
    const toks = estimateTokens(text);
    parts.push(text);
    extraParts.push(text);
    used += toks;
    tauTyped += toks;
    packedNorm.add(key);
    typedAdded = true;
  }

  let chunkAdded = false;
  for (const chunk of args.ranked_chunks ?? []) {
    const text = (chunk || "").trim();
    if (!text) continue;
    const key = text.toLowerCase();
    if (packedNorm.has(key) || containedInHot(text, hot)) continue;
    if (blocked(text)) {
      dupSuppressed += estimateTokens(text);
      continue;
    }
    if (marginal(text)) {
      dupSuppressed += estimateTokens(text);
      continue;
    }
    if (!fits(text)) continue;
    const toks = estimateTokens(text);
    parts.push(text);
    extraParts.push(text);
    used += toks;
    tauRanked += toks;
    packedNorm.add(key);
    chunkAdded = true;
  }

  let spanAdded = false;
  for (const chunk of args.span_chunks ?? []) {
    const text = (chunk || "").trim();
    if (!text) continue;
    const key = text.toLowerCase();
    if (packedNorm.has(key) || containedInHot(text, hot)) continue;
    if (blocked(text)) {
      dupSuppressed += estimateTokens(text);
      continue;
    }
    if (marginal(text)) {
      dupSuppressed += estimateTokens(text);
      continue;
    }
    if (!fits(text)) continue;
    const toks = estimateTokens(text);
    parts.push(text);
    extraParts.push(text);
    used += toks;
    tauSpans += toks;
    packedNorm.add(key);
    spanAdded = true;
  }

  let body = parts.join("\n\n").trim();
  let { ws, chars4 } = countTokens(body);
  let packed = chars4;
  if (packed > cap && body) {
    body = truncateToBudget(body, cap);
    ({ ws, chars4 } = countTokens(body));
    packed = chars4;
  }
  if (typedAdded || chunkAdded || spanAdded) method = "query-pack";
  else if (!hot) method = chunkAdded || spanAdded ? "extractive" : "hot_set";

  const skip =
    Boolean(args.allow_skip) &&
    crossTurnDedupEnabled() &&
    !args.openitem_changed &&
    !args.node_superseded &&
    packed < skipFloor;
  if (skip) {
    return {
      text: "",
      packed_tokens: 0,
      tokens_ws: 0,
      tokens_chars4: 0,
      budget: cap,
      rate: 0,
      method: "skip",
      novel_tokens: 0,
      dup_suppressed_tokens: dupSuppressed + packed,
      line_hashes: [],
      tau_hot: 0,
      tau_typed: 0,
      tau_ranked: 0,
      tau_spans: 0,
    };
  }

  const hashes = body
    .split("\n")
    .filter((l) => l.trim())
    .map((l) => lineHash(l));
  let novel = packed;
  if (suppress.size && hashes.length) {
    const novelParts = body
      .split("\n")
      .filter((line) => line.trim() && !args.recent_hashes?.has(lineHash(line)));
    novel = novelParts.length ? estimateTokens(novelParts.join("\n")) : 0;
  }
  return {
    text: body,
    packed_tokens: packed,
    tokens_ws: ws,
    tokens_chars4: chars4,
    budget: cap,
    rate: cap ? packed / cap : 0,
    method,
    novel_tokens: novel,
    dup_suppressed_tokens: dupSuppressed,
    line_hashes: hashes,
    tau_hot: tauHot,
    tau_typed: tauTyped,
    tau_ranked: tauRanked,
    tau_spans: tauSpans,
  };
}

function containedInHot(text: string, hot: string): boolean {
  if (!hot || !text) return false;
  const needle = text.toLowerCase();
  const hay = hot.toLowerCase();
  if (hay.includes(needle)) return true;
  if (text.includes(":")) {
    const rest = text.split(":", 2)[1]?.trim().toLowerCase() ?? "";
    if (rest && hay.includes(rest)) return true;
  }
  return false;
}

function truncateToBudget(text: string, budget: number): string {
  if (estimateTokens(text) <= budget) return text;
  const maxChars = Math.max(1, budget * 4);
  if ([...text].length <= maxChars) return text;
  // Prefer code-point truncate to match estimateTokens
  const chars = [...text];
  return chars.slice(0, Math.max(0, maxChars - 3)).join("").replace(/\s+$/, "") + "...";
}

File: .clawhub-stage/package/engine/src/chat_compressor/pack.py
"""Token-budget packer: HOT_SET → typed lines → ranked chunks."""

from __future__ import annotations

import hashlib
import os
from dataclasses import dataclass, field

from chat_compressor.extractive import jaccard, keyword_set
from chat_compressor.metrics import estimate_tokens


DEFAULT_FORWARD_BUDGET = 2048
DEDUP_K = 3
WARMUP_TURNS = 3
# recall-0.5 defaults (cursor-parity callers may pass 0.8 / 0.5 explicitly)
MARGINAL_JACCARD = 0.92
SKIP_FLOOR_TOKENS = 64
NOVELTY_BUDGET_FLOOR = 1.0


def forward_budget() -> int:
    raw = os.environ.get("CHAT_COMPRESSOR_FORWARD_BUDGET", "").strip()
    if not raw:
        return DEFAULT_FORWARD_BUDGET
    try:
        return max(1, int(raw))
    except ValueError:
        return DEFAULT_FORWARD_BUDGET


def novelty_budget_floor() -> float:
    raw = os.environ.get("CHAT_COMPRESSOR_NOVELTY_FLOOR", "").strip()
    if not raw:
        return float(NOVELTY_BUDGET_FLOOR)
    try:
        return max(0.0, min(1.0, float(raw)))
    except ValueError:
        return float(NOVELTY_BUDGET_FLOOR)


def marginal_jaccard_threshold(default: float = MARGINAL_JACCARD) -> float:
    raw = os.environ.get("CHAT_COMPRESSOR_MARGINAL_JACCARD", "").strip()
    if not raw:
        return float(default)
    try:
        return max(0.0, min(1.0, float(raw)))
    except ValueError:
        return float(default)


def cross_turn_dedup_enabled() -> bool:
    raw = os.environ.get("CHAT_COMPRESSOR_CROSS_TURN_DEDUP", "1").strip().lower()
    return raw not in {"0", "false", "no", "off"}


def count_tokens(text: str) -> tuple[int, int]:
    """Return (whitespace_tokens, chars/4_tokens)."""
    ws = len(text.split()) if text.strip() else 0
    chars4 = estimate_tokens(text) if text else 0
    return ws, chars4


def line_hash(text: str) -> str:
    return hashlib.sha1(text.strip().lower().encode("utf-8")).hexdigest()[:16]


def adaptive_budget(
    t: int,
    novelty_rate: float,
    cap: int | None = None,
    floor: float | None = None,
) -> int:
    """B_max for the first 3 turns, then scale to rolling novelty (capped)."""
    limit = int(cap if cap is not None else forward_budget())
    limit = max(1, limit)
    if int(t) <= WARMUP_TURNS:
        return limit
    rate = max(0.0, min(1.0, float(novelty_rate)))
    rho_min = novelty_budget_floor() if floor is None else float(floor)
    scaled = int(limit * max(rho_min, rate))
    return max(SKIP_FLOOR_TOKENS, min(limit, scaled))


@dataclass(frozen=True)
class PackResult:
    text: str
    packed_tokens: int
    tokens_ws: int
    tokens_chars4: int
    budget: int
    rate: float
    method: str
    novel_tokens: int = 0
    dup_suppressed_tokens: int = 0
    line_hashes: tuple[str, ...] = field(default_factory=tuple)
    tau_hot: int = 0
    tau_typed: int = 0
    tau_ranked: int = 0
    tau_spans: int = 0


def pack_forward(
    *,
    hot_set: str = "",
    typed_lines: list[str] | None = None,
    ranked_chunks: list[str] | None = None,
    span_chunks: list[str] | None = None,
    budget: int | None = None,
    recent_hashes: set[str] | None = None,
    openitem_changed: bool = True,
    node_superseded: bool = False,
    allow_skip: bool = False,
    marginal_jaccard: float | None = None,
    skip_floor_tokens: int = SKIP_FLOOR_TOKENS,
) -> PackResult:
    """Pack in order. HOT_SET → typed → ranked → spans. Cap at budget (chars/4)."""
    cap = int(budget if budget is not None else forward_budget())
    cap = max(1, cap)
    mu = marginal_jaccard_threshold() if marginal_jaccard is None else float(marginal_jaccard)
    parts: list[str] = []
    used = 0
    packed_norm: set[str] = set()
    method = "hot_set"
    dup_suppressed = 0
    tau_hot = 0
    tau_typed = 0
    tau_ranked = 0
    tau_spans = 0
    suppress = set(recent_hashes or ())
    if node_superseded or not cross_turn_dedup_enabled():
        suppress = set()

    def _blocked(text: str) -> bool:
        if not suppress:
            return False
        return line_hash(text) in suppress

    extra_parts: list[str] = []

    def _marginal(text: str) -> bool:
        if not extra_parts:
            return False
        body = "\n".join(extra_parts)
        return jaccard(keyword_set(text), keyword_set(body)) > mu

    hot = (hot_set or "").strip()
    if hot:
        piece = f"HOT_SET:\n{hot}"
        toks = estimate_tokens(piece)
        if toks > cap:
            piece = _truncate_to_budget(piece, cap)
            toks = estimate_tokens(piece)
        parts.append(piece)
        used += toks
        tau_hot = toks
        packed_norm.add(hot.lower())
        for line in hot.splitlines():
            packed_norm.add(line.strip().lower())

    def _fits(block: str) -> bool:
        return used + estimate_tokens(block) <= cap

    typed_added = False
    for line in typed_lines or []:
        text = (line or "").strip()
        if not text:
            continue
        key = text.lower()
        if key in packed_norm:
            continue
        if _blocked(text):
            dup_suppressed += estimate_tokens(text)
            continue
        if _marginal(text):
            dup_suppressed += estimate_tokens(text)
            continue
        if not _fits(text):
            continue
        toks = estimate_tokens(text)
        parts.append(text)
        extra_parts.append(text)
        used += toks
        tau_typed += toks
        packed_norm.add(key)
        typed_added = True

    chunk_added = False
    for chunk in ranked_chunks or []:
        text = (chunk or "").strip()
        if not text:
            continue
        key = text.lower()
        if key in packed_norm or _contained_in_hot(text, hot):
            continue
        if _blocked(text):
            dup_suppressed += estimate_tokens(text)
            continue
        if _marginal(text):
            dup_suppressed += estimate_tokens(text)
            continue
        if not _fits(text):
            continue
        toks = estimate_tokens(text)
        parts.append(text)
        extra_parts.append(text)
        used += toks
        tau_ranked += toks
        packed_norm.add(key)
        chunk_added = True

    span_added = False
    for chunk in span_chunks or []:
        text = (chunk or "").strip()
        if not text:
            continue
        key = text.lower()
        if key in packed_norm or _contained_in_hot(text, hot):
            continue
        if _blocked(text):
            dup_suppressed += estimate_tokens(text)
            continue
        if _marginal(text):
            dup_suppressed += estimate_tokens(text)
            continue
        if not _fits(text):
            continue
        toks = estimate_tokens(text)
        parts.append(text)
        extra_parts.append(text)
        used += toks
        tau_spans += toks
        packed_norm.add(key)
        span_added = True

    body = "\n\n".join(parts).strip()
    ws, chars4 = count_tokens(body)
    packed = chars4
    if packed > cap and body:
        body = _truncate_to_budget(body, cap)
        ws, chars4 = count_tokens(body)
        packed = chars4
    if typed_added or chunk_added or span_added:
        method = "query-pack"
    elif not hot:
        method = "extractive" if (chunk_added or span_added) else "hot_set"

    skip = (
        allow_skip
        and cross_turn_dedup_enabled()
        and not openitem_changed
        and not node_superseded
        and packed < skip_floor_tokens
    )
    if skip:
        hashes: tuple[str, ...] = ()
        return PackResult(
            text="",
            packed_tokens=0,
            tokens_ws=0,
            tokens_chars4=0,
            budget=cap,
            rate=0.0,
            method="skip",
            novel_tokens=0,
            dup_suppressed_tokens=dup_suppressed + packed,
            line_hashes=hashes,
            tau_hot=0,
            tau_typed=0,
            tau_ranked=0,
            tau_spans=0,
        )

    hashes = tuple(line_hash(line) for line in body.splitlines() if line.strip())
    novel = packed
    if suppress and hashes:
        novel_body_parts = [
            line
            for line in body.splitlines()
            if line.strip() and line_hash(line) not in (recent_hashes or set())
        ]
        novel = estimate_tokens("\n".join(novel_body_parts)) if novel_body_parts else 0
    rate = (packed / cap) if cap else 0.0
    return PackResult(
        text=body,
        packed_tokens=packed,
        tokens_ws=ws,
        tokens_chars4=chars4,
        budget=cap,
        rate=rate,
        method=method,
        novel_tokens=novel,
        dup_suppressed_tokens=dup_suppressed,
        line_hashes=hashes,
        tau_hot=tau_hot,
        tau_typed=tau_typed,
        tau_ranked=tau_ranked,
        tau_spans=tau_spans,
    )


def _contained_in_hot(text: str, hot: str) -> bool:
    if not hot or not text:
        return False
    needle = text.lower()
    hay = hot.lower()
    if needle in hay:
        return True
    # typed "OpenItem: milk" vs hot "OpenItem milk: open: milk"
    if ":" in text:
        rest = text.split(":", 1)[1].strip().lower()
        if rest and rest in hay:
            return True
    return False


def _truncate_to_budget(text: str, budget: int) -> str:
    if estimate_tokens(text) <= budget:
        return text
    # chars/4 ≈ budget → max_chars = budget * 4
    max_chars = max(1, budget * 4)
    if len(text) <= max_chars:
        return text
    return text[: max(0, max_chars - 3)].rstrip() + "..."
